silo-to-silo/contracts/src/SiloToSilo.sol

```diff
@@ -0,0 +1,391 @@
+        string memory _siloAccountId,
+        string memory _nativeTokenAccountId
+    ) external initializer {
+        near = AuroraSdk.initNear(IERC20_NEAR(wnearAddress));
+        siloAccountId = _siloAccountId;
+        registeredTokens[IEvmErc20(address(0))].nearTokenAccountId = _nativeTokenAccountId;
+
+        _grantRole(CALLBACK_ROLE, AuroraSdk.nearRepresentitiveImplicitAddress(address(this)));
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(PAUSE_ADMIN_ROLE, msg.sender);
+        _grantRole(UNPAUSE_ADMIN_ROLE, msg.sender);
+    }
+
+    function registerToken(IEvmErc20 token) external whenNotPaused {
+        require(bytes(registeredTokens[token].nearTokenAccountId).length == 0, "The token is already registered");
+
+        PromiseCreateArgs memory callGetNep141FromErc20 = near.call(
+            siloAccountId,
+            "get_nep141_from_erc20",
+            abi.encodePacked(address(token)),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(this.getNep141FromErc20Callback.selector, token),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+
+        callGetNep141FromErc20.then(callback).transact();
+    }
+
+    function getNep141FromErc20Callback(IEvmErc20 token) external onlyRole(CALLBACK_ROLE) {
+        require(
+            AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful,
+            "ERROR: The `get_nep141_from_erc20()` XCC call failed"
+        );
+
+        string memory nearTokenAccountId = string(AuroraSdk.promiseResult(0).output);
+        registeredTokens[token].nearTokenAccountId = nearTokenAccountId;
+        emit TokenRegistered(token, nearTokenAccountId);
+    }
+
+    function storageDeposit(IEvmErc20 token, uint128 storageDepositAmount) external whenNotPaused {
+        require(registeredTokens[token].isStorageRegistered == false, "The storage is already registered");
+        _storageDeposit(token, storageDepositAmount, getImplicitNearAccountIdForSelf(), false);
+    }
+
+    function recipientStorageDeposit(
+        IEvmErc20 token,
+        uint128 storageDepositAmount,
+        string memory recipientId
+    ) external whenNotPaused {
+        require(registeredRecipients[token][recipientId] == false, "The storage is already registered");
+        _storageDeposit(token, storageDepositAmount, recipientId, true);
+    }
+
+    function _storageDeposit(
+        IEvmErc20 token,
+        uint128 storageDepositAmount,
+        string memory recipientId,
+        bool isRecipientStorageDeposit
+    ) private {
+        string memory nearTokenAccountId = registeredTokens[token].nearTokenAccountId;
+        require(bytes(nearTokenAccountId).length > 0, "The token is not registered");
+
+        PromiseCreateArgs memory callStorageDeposit = near.call(
+            nearTokenAccountId,
+            "storage_deposit",
```

Comment:
Do I understand correctly that for this to work you need to transfer tokens to the SiloToSiloImplicitNearAccountId? And we don't have any refund mechanism?

Reply:
cb8ff916385fd4675cb43bb4b95ded002e7bc44c